src/intrinsics/ecma262/Object.js

```diff
@@ -36,11 +37,166 @@ import {
   SetIntegrityLevel,
   HasSomeCompatibleType,
 } from "../../methods/index.js";
-import { Create, Properties as Props, To } from "../../singletons.js";
+import { Create, Havoc, Properties as Props, To } from "../../singletons.js";
 import type { BabelNodeExpression } from "babel-types";
 import * as t from "babel-types";
 import invariant from "../../invariant.js";
 
+function handleObjectAssignSnapshot(
+  to: ObjectValue | AbstractObjectValue,
+  frm: ObjectValue | AbstractObjectValue,
+  frm_was_partial: boolean,
+  delayedSources: Array<Value>
+): void {
+  if (to instanceof AbstractObjectValue && to.values.isTop()) {
+    // We don't know which objects to make partial and making all objects partial is failure in itself
+    AbstractValue.reportIntrospectionError(to);
+    throw new FatalError();
+  } else {
+    // if to has properties, we better remove them because after the temporal call to Object.assign we don't know their values anymore
+    if (to.hasStringOrSymbolProperties()) {
+      // preserve them in a snapshot and add the snapshot to the sources
+      delayedSources.push(to.getSnapshot({ removeProperties: true }));
+    }
+
+    if (frm instanceof ObjectValue && frm.mightBeHavocedObject()) {
+      // it's not safe to trust any of its values
+      delayedSources.push(frm);
+    } else if (frm_was_partial) {
+      if (frm instanceof AbstractObjectValue && frm.kind === "explicit conversion to object") {
+        // Make it implicit again since it is getting delayed into an Object.assign call.
+        delayedSources.push(frm.args[0]);
+      } else {
+        let frmSnapshot = frm.getSnapshot();
+        frm.temporalAlias = frmSnapshot;
+        frm.makePartial();
+        delayedSources.push(frmSnapshot);
+      }
+    }
```

Comment:
This doesn't look fixed yet. https://github.com/facebook/prepack/pull/2068/#discussion_r196282693

It is possible to fall into this else branch (not havoced and not partial), in the error recovery scenario and that will cause this object to be completely ignored. One possible solution is to havoc it before calling this.

Reply:
We can't havoc it before going into this, or we can't get a snapshot from it properly. I've changed to code to push it to delayedSources.